ignite/engine/engine.py

```diff
@@ -706,23 +708,23 @@ def switch_batch(engine):
         return self._internal_run()
 
     @staticmethod
-    def _init_timers(state: State):
+    def _init_timers(state: State) -> None:
         state.times[Events.EPOCH_COMPLETED.name] = 0.0
         state.times[Events.COMPLETED.name] = 0.0
 
-    def _get_data_length(self, data):
-        data_length = None
+    def _get_data_length(self, data: Optional[Iterable] = None) -> Optional[int]:
         try:
             if hasattr(data, "__len__"):
-                data_length = len(data)
+                return len(data)  # type: ignore[arg-type]
         except TypeError:
             # _InfiniteConstantSampler can raise a TypeError on DataLoader length of a IterableDataset
             pass
-        return data_length
+        return None
 
     def _setup_engine(self) -> None:
         iteration = self.state.iteration
-        self._dataloader_iter = iter(self.state.dataloader)
+        if self.state.dataloader is not None:
```

Comment:
I suppose that this check is related to the fact that this function can be used with default `self.state = State()` and where dataloader is None. Let's raise an exception (as above) if it is None instead.

Reply:
ok, gotcha 👍